File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if( root == NULL )
            return;
        TreeNode *temp = root;
        root = temp;
        while(temp)
        {
            if(temp->left!=NULL)
            {
                TreeNode* pre = temp->left;
                while(pre->right)
                    pre=pre->right;
                pre->right = temp->right;
                temp->right=temp->left;
                temp->left = NULL;
            }
            temp=temp->right;
        }
    }
};

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        ArrayList<TreeNode> list = new ArrayList<>();
        
        preorder(list,root);
        TreeNode pre = root;
        TreeNode curr = null;
        for(int i =1;i<list.size();i++) {
            curr = list.get(i);
            pre.left = null;
            pre.right = curr;
            pre = curr;
        }
    }
    public void preorder(ArrayList<TreeNode> list ,TreeNode root) {
        if (root == null) {
            return;
        }
        list.add(root);
        preorder(list,root.left);
        preorder(list,root.right);
    }
}